src/physics/em/detail/UrbanMscScatter.hh

```diff
@@ -505,29 +469,73 @@ CELER_FUNCTION real_type UrbanMscScatter::calc_correction(real_type tau) const
 //---------------------------------------------------------------------------//
 /*!
  * Sample the displacement direction using G4UrbanMscModel::SampleDisplacement
- * of Geant4 10.7: simple and fast sampling based on single scattering results.
+ * (simple and fast sampling based on single scattering results) and update
+ * direction and position of the particle.
  *
  * A simple distribution for the unit direction on the lateral (x-y) plane,
  * \f$ Phi = \phi \pm \psi \f$ where \f$ psi \sim \exp(-\beta*v) \f$ and
  * \f$\beta\f$ is determined from the requirement that the distribution should
  * give the same mean value that is obtained from the single scattering
  * simulation.
  *
- * @param phi the azimuthal angle of the multiple scattering.
+ * @param cth the cos(\theta) of the multiple scattering.
+ * @param rmax2 the asymmetry between true path and geom path
  */
 template<class Engine>
-CELER_FUNCTION Real3 UrbanMscScatter::dir_displacement(Engine&   rng,
-                                                       real_type phi) const
+CELER_FUNCTION void
+UrbanMscScatter::update_geometry(Engine& rng, real_type cth, real_type rmax2)
 {
-    constexpr real_type cbeta  = 2.160;
-    real_type           cbeta1 = 1 - std::exp(-cbeta * constants::pi);
+    // Update direction
+    UniformRealDistribution<real_type> sample_phi(0, 2 * constants::pi);
+    real_type                          phi = sample_phi(rng);
+    Real3 direction = rotate(from_spherical(cth, phi), inc_direction_);
+
+    geometry_.set_dir(direction);
```

Comment:
Instead of modifying the geometry state in the bowels of the MSC scattering, could you have the MSC scatter result return an updated direction and a displacement? (Represent the result as a change in state rather than changing the state here.) Since the geometry state isn't always changed, we will want to add flags for direction/position needing update, but I think rather than worry about that particular feature now I'll be incorporating it into the "action" refactor.

Reply:
Rolled back to the previous version and moved the geometry update to inside LDemoLauncher::AlongAndPostStepLauncher.